### PR TITLE
Expose build mode in environment of asset transformer processes

### PR DESCRIPTION
**Pull request by andrewkolos**
In service of https://github.com/flutter/flutter/issues/143348

When invoking a package to transform an asset, we set `FLUTTER_BUILD_MODE` to the CLI name of the build mode being used. Inspired by https://github.com/flutter/flutter/issues/101077#issuecomment-1890379501:

> Do transformers know whether they get executed in debug or release mode? I kinda imagine that being useful. Ex: There's a transformer that optimizes the file size of images. Depending on the amount and size of the images, that could take a significant amount of time. Therefore, I might want to only execute it in release builds.

Note for the reviewer: the interesting part of this change can be found in the commit [set environment variable to build mode when running asset transformer…](https://github.com/flutter/flutter/pull/144752/commits/579912d470b6f6eb036fd6b23a128fa3942f6282). The rest of the change is updating call sites with a new argument.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart [hunk 1]
```diff
@@ -22,13 +23,18 @@ final class AssetTransformer {
...
     required FileSystem fileSystem,
     required String dartBinaryPath,
+    required BuildMode buildMode,
   })  : _processManager = processManager,
         _fileSystem = fileSystem,
-        _dartBinaryPath = dartBinaryPath;
+        _dartBinaryPath = dartBinaryPath,
+        _buildMode = buildMode;
+
+  static const String buildModeEnvVar = 'FLUTTER_BUILD_MODE';
```
Comment (andrewkolos) [inline packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:32]: I have an internal instinct telling me that these should be namespaced, but `FLUTTER_ASSET_TRANFORMER__FLUTTER_BUILD_MODE` feels a bit verbose. I think we are fine here.

**Comment by andrewkolos** [comment]
I do think a _better_ solution would be to allow users to define separate asset configurations per build mode. For purely demonstrative purposes, consider this example fictional pubspec:

```yaml
flutter:
  assets:
    - path: assets/hello.png
      buildModes: ["debug"]
    - path: assets/hello.png
      buildModes: ["profile", "release"]
      transformers:
        - package: optimizer
          args: --quality 95
```

However, this would be 1) harder to implement and maintain 2) would call for a bunch of design discussion. I think this PR is still useful as a stop gap. I imagine, for example, an image optimization transformer could provide an option like "--release-only" which tells the transformer to just copy the file as-is when not running in release mode.

```yaml
flutter:
  assets:
    - path: assets/hello.png
      transformers:
        - package: optimizer
          args: --release-only --quality 95
```

File: packages/flutter_tools/lib/src/build_system/targets/assets.dart [hunk 2]
```diff
@@ -343,6 +345,11 @@ class CopyAssets extends Target {
 
   @override
   Future<void> build(Environment environment) async {
+    final String? buildModeEnvironment = environment.defines[kBuildMode];
+    if (buildModeEnvironment == null) {
+      throw MissingDefineException(kBuildMode, name);
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_system/targets/assets.dart:350]: i wonder if we'll ever hit this. merging this and then seeing if we need to roll back sounds good to me, though.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Comment by andrewkolos** [comment]
Reason for revert: compilation issue has turned the tree red

**Comment by christopherfujino** [comment]
> Reason for revert: compilation issue has turned the tree red

For posterity can you link a failing a build?

**Comment by andrewkolos** [comment]
> > Reason for revert: compilation issue has turned the tree red
> 
> For posterity can you link a failing a build? (This is something I always check when reviewing re-lands)

https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8753729103951007409/+/u/flutter_config_--clear-features/stdout

```
Building flutter tool...
Resolving dependencies...
Got dependencies.
flutter/packages/flutter_tools/lib/src/bundle_builder.dart:177:61: Error: Required named parameter 'buildMode' must be provided.
  final AssetTransformer assetTransformer = AssetTransformer(
                                                            ^
flutter/packages/flutter_tools/lib/src/build_system/tools/asset_transformer.dart:22:3: Context: Found this candidate, but the arguments don't match.
  AssetTransformer({
```

**Comment by Jasguerrero** [comment]
@CaseyHillers is it possible to get the presubmit google testing run for this PR? I see the check passed but it failed postsubmit

**Comment by andrewkolos** [comment]
> @CaseyHillers is it possible to get the presubmit google testing run for this PR? I see the check passed but it failed postsubmit

The PR branch was stale—it branched off master before https://github.com/flutter/flutter/pull/144734 landed. _That_ PR introduced a new `AssetTransformer` call site, while _this_ PR modified the signature of `AssetTransformer`. I'm assuming this is why the checks initially passed.

**Comment by Jasguerrero** [comment]
> The PR branch was stale—it branched off master before #144734 landed. _That_ PR introduced a new `AssetTransformer` call site, while _this_ PR modified the signature of `AssetTransformer`. I'm assuming this is why the checks initially passed.

Interesting. Thanks for the explanation @andrewkolos!

**Pull request merged**